overlord/snapstate/snapstate.go

```diff
@@ -4146,6 +4148,120 @@ func GadgetConnections(st *state.State, deviceCtx DeviceContext) ([]gadget.Conne
 	return gadgetInfo.Connections, nil
 }
 
+// downloadsToKeep returns a map of download file names that need to be kept
+// for all current snaps in the system state.
+//
+// A downloaded file is only kept if any of the following are true:
+//    1. The revision is in SnapState.Sequence
+//    2. The revision is in saved in refresh-candidates
+//    3. The revision is pointed to by a download task in an ongoing change
+//
+// It is the caller's responsibility to lock state before calling this function.
+func downloadsToKeep(st *state.State) (map[string]bool, error) {
+	snapStates, err := All(st)
+	if err != nil {
+		return nil, err
+	}
+
+	// initialize refresh hints
+	var refreshHints map[string]*refreshCandidate
+	if err := st.Get("refresh-candidates", &refreshHints); err != nil {
+		if !errors.Is(err, &state.NoStateError{}) {
+			return nil, fmt.Errorf("cannot get refresh-candidates: %v", err)
+		}
+	}
+
+	downloadsToKeep := map[string]bool{}
+	keep := func(name string, rev snap.Revision) {
+		downloadsToKeep[fmt.Sprintf("%s_%s.snap", name, rev)] = true
+	}
```

Comment:
Not a huge deal but I'd create the map only if keep is called and return nil otherwise. The callers don't even have to change